### ISSUE TITLE
Remove song from playlist

### ISSUE THREAD

**Issue by starchyunderscore** [OxygenCobalt/Auxio#624]
### Description

An option to remove a song from a playlist, possibly in the meatball menu of the song in the playlist, or by clicking it again in the "add to playlist" menu.

### Problem solved

I accidentally added a song to the playlist twice. I trid to fix it by un-checking the song in the "add to playlist" menu. It is now in the playlist five times.

### Other implementations

my only other experience is with spotify. It has a second "remove from playlist" menu.

### Benefit

I can fix my mistakes without discarding the whole playlist and starting over.

### Duplicates

- [X] I have checked this feature request for duplicates.
- [X] I have checked that this feature is not implemented in the [lastest version](https://github.com/OxygenCobalt/Auxio/releases).
- [X] I have checked the [Why Are These Features Missing?](https://github.com/OxygenCobalt/Auxio/wiki/Why-Are-These-Features-Missing%3F) page.
- [X] I agree to the [Contribution Guidelines](https://github.com/OxygenCobalt/Auxio/blob/dev/.github/CONTRIBUTING.md).

**Comment by OxygenCobalt** [comment]
You can remove songs by pressing the edit button in the playlist view and then swiping the song you want deleted away. Albeit, since this is like the 4th time that someone has requested this, I'm just going to add an option to the menu.

**Comment by Chzy2018** [comment]
In the current interface of the playlist, there is no indication of how to delete a song after pressing the edit button and swiping, which may lead to some users, including myself, being unable to locate the option to remove songs from the playlist. It is recommended that a removal button be added next to each song in the playlist to facilitate this process. 

Additionally, I'd like to inquire whether the ability to add the same song multiple times to a single playlist is a bug in the software. This can result in duplication of certain tracks within the same playlist if I am unsure whether they have already been added. I would greatly appreciate it if the software could verify whether a song is already in the playlist before it is added. Thank you for your efforts, an excellent software!

**Comment by e-zk** [comment]
> I would greatly appreciate it if the software could verify whether a song is already in the playlist before it is added.

FWIW, if you go to add a song to a playlist the playlist preview will have a "tick" on it to show that the song is already in there.

**Comment by Chzy2018** [comment]
> FWIW, if you go to add a song to a playlist the playlist preview will have a "tick" on it to show that the song is already in there.

Thank you for your response. I have conducted an experiment with adding individual songs and noted this feature. 

However, I typically select and add multiple songs or entire albums to a playlist simultaneously. When some of these songs are already in the playlist and others are not, the playlist icon does not show the "tick", making it difficult for me to discern which songs have already been added. This situation often leads to repeated tracks in my playlist, which is frustrating. Therefore, I am in need of a feature that would prevent duplicate songs from being added to the playlist. I hope this functionality can be implemented.

**Comment by OxygenCobalt** [comment]
> Additionally, I'd like to inquire whether the ability to add the same song multiple times to a single playlist is a bug in the software. This can result in duplication of certain tracks within the same playlist if I am unsure whether they have already been added. I would greatly appreciate it if the software could verify whether a song is already in the playlist before it is added. Thank you for your efforts, an excellent software!

Yeah, you can add duplicate songs to a playlist. This is intentional design since perhaps some people want to deliberately add duplicates to increase the chance it shows up in shuffles. 

I do want to add playlist deduplication eventually with #451.

> However, I typically select and add multiple songs or entire albums to a playlist simultaneously. When some of these songs are already in the playlist and others are not, the playlist icon does not show the "tick", making it difficult for me to discern which songs have already been added. This situation often leads to repeated tracks in my playlist, which is frustrating. Therefore, I am in need of a feature that would prevent duplicate songs from being added to the playlist. I hope this functionality can be implemented.

I don't know how to best indicate "All songs are already in this playlist" vs. "*Some* songs are already in this playlist" A different type of checkmark would not be clear enough. Any ideas, perhaps @Chzy2018?

**Comment by Chzy2018** [comment]
> I don't know how to best indicate "All songs are already in this playlist" vs. "_Some_ songs are already in this playlist" A different type of checkmark would not be clear enough. Any ideas, perhaps @Chzy2018?

I propose implementing a secondary confirmation feature when songs are added to a playlist. If the song already exists in the playlist, a prompt could appear stating, "The following songs are already in the playlist. Would you like to add them again?" with options to "Add all songs" or "Add only those songs not currently in the playlist." If all the selected songs are new to the playlist, they could be added without any prompt. Is such a feature implementation feasible? @OxygenCobalt

**Comment by OxygenCobalt** [comment]
That works, but the issue lies in that it's a very wordy and nuanced thing to specify, so I would have to think more about how to implement it.

**Comment by bad1080** [comment]
> You can remove songs by pressing the edit button in the playlist view and then swiping the song you want deleted away. Albeit, since this is the 4th time that someone has requested this, I'm just going to add an option to the menu to reduce confusion.

this requires you to scroll all the way up to the edit button and then back down to where the song was, it is highly unpractical with larger playlists

**Comment by OxygenCobalt** [comment]
Okay, so I can do this, but there's a caveat in that this will *not* work on playlists with several songs in them. I've yet to really build good state management for handling duplicate songs since it's extremely hard to get right. As a result, I think any kind of "delete from playlist" action will either:

1. Delete all instances of the song from the playlist
2. Fail when duplicates are in the playlist

What do you all like more @bad1080 @Chzy2018 @ez-k

**Comment by bad1080** [comment]
> Okay, so I can do this, but there's a caveat in that this will _not_ work on playlists with several songs in them. I've yet to really build good state management for handling duplicate songs since it's extremely hard to get right. As a result, I think any kind of "delete from playlist" action will either:
> 
> 1. Delete all instances of the song from the playlist
> 2. Fail when duplicates are in the playlist
> 
> What do you all like more @bad1080 @Chzy2018 @Ez-K

i have a playlist with over 27k songs on it so scrolling to the top of the list to remove something is just not possible for me. so since i am not using it for duplicates i don't care if all of them are gone, so option no. 1 for me!
i just realized i'd need the ability to do just that but from the "queue" since removing it from the playlist involves several more steps (and finding the song i want to remove on the playlist would become equally impossible)... but i guess that wont be possible with the way the playlist is converted to a queue once you start playing it : (

**Comment by gummyhulk** [comment]
I'm not sure if anyone else mentioned, but would it be possible to add "delete from playlist" option in the 3 dots beside the songs? Like in Spotify?

**Comment by OxygenCobalt** [comment]
Yes, this is the whole point of the issue @gummyhulk

**Comment by OxygenCobalt** [comment]
So, I think what I need to do is:
1. Forbid duplicates in a playlist (External problem)
2. Uniquely identify individual entries within a playlist, even when duplicate (Internal problem)

I think I can do 2 with more refactors, so I guess that's the aim. Does mean I'll prioritize other stuff for now.

**Comment by OxygenCobalt** [comment]
Yeah, I need to introduce a new construct into the music model that indicates some extra differentiation in playlist songs for this to be feasible, will do that.

**Comment by disbelief2389** [comment]
This issue has been bugging me personally.
The fact that I had to come to this GitHub with the intent to fix the issue myself, to figure out that the way to do it is by clicking edit in the playlist view, is a serious design flaw.
Thus, I will be making a version that works and submitting a pull request for this, even though I'm a pretty novice programmer.
You should at least have an option to do so in more than one place, especially when the place is so evidently discrete.

**Comment by Mincan2** [comment]
This is currently a horrible way to delete a song you are currently listening to. There is no way to know where in the playlist it is located to remove it. Please make it much easier to remove songs from a playlist either in the hamburger menu or create a favoured favourites list that you can easily add/remove with an icon like in Samsung Music with the heart.

basically it is impossible to remove songs from playlists if they are of any size. My favourites playlist is over 600 songs.

**Comment by OxygenCobalt** [comment]
Yeah I'm sorry @Mincan2 @disbelief2389 for not doing this for so long. I keep getting hung up on the technical edge cases which is really stupid. Will work on it

**Comment by disbelief2389** [comment]
> Yeah I'm sorry @Mincan2 @disbelief2389 for not doing this for so long. I keep getting hung up on the technical edge cases which is really stupid. Will work on it

You're allgood. If I could get the environment setup locally I could probably get it done but like I said I am a novice, but also am just busy.

That said it would be good to implement this as well as a volume slider for Auxio as a whole. Add it in the audio settings alongside the equaliser etc. Because personally my phone doesnt allow per app volume, but for my workflows it is quite necessary.

Anyway thanks for your attention.

**Comment by nazar-pc** [comment]
I spent like 10 minutes yesterday and was unable to remove tracks from stored playlists or the current queue. I tried swiping, I looked at all menus and was very frustrated.

Was about to report that this obvious feature is missing just now, but turns out I was swiping right, but needed to swipe left instead 😅
It should be added to the context menus and maybe made more discoverable somehow.

**Comment by OxygenCobalt** [comment]
I know, I intend to but my bandwidth is often limited by work / not trying to overwork myself because of said work. Sorry @nazar-pc  I want to get to it soon

**Comment by Mincan2** [comment]
Any updates or estimates of when this feature will be added? 
I might have some time in September to implement the feature myself. What is the preferred mechanism for doing so? Like should it be an additional option in the songs meatball menu or clicking the playlist its already part of in the add to playlist feature...
